### PR TITLE
Batch-wise reduce-sum layer

### PR DESCRIPTION
**Pull request by timmoon10**
Some of our applications require batch-wise metrics, e.g. R2 over a mini-batch. In principle, we can implement a lot of these with a sum that goes over the batch dimension (e.g. `torch.sum(x, dim=0)` in PyTorch). For example, if we want to compute the (biased) variance of a scalar over a mini-batch, we can now do:
```
# var(x) = mean(x^2) - mean(x)^2
mini_batch_size = lbann.MiniBatchSize()
mean = lbann.Divide(lbann.BatchwiseReduceSum(x), mini_batch_size)
sqmean = lbann.Divide(lbann.BatchwiseReduceSum(lbann.Square(x)), mini_batch_size)
var = lbann.Subtract(sqmean, lbann.Square(mean))

# var(x) = mean( (x-mean(x))^2 )
# Note: This is better than the above for numerical accuracy
mini_batch_size = lbann.MiniBatchSize()
mean = lbann.Divide(lbann.BatchwiseReduceSum(x), mini_batch_size)
var = lbann.Divide(lbann.BatchwiseReduceSum(lbann.Square(lbann.Subtract(x, mean))), mini_batch_size)
```
Note that this layer requires blocking communication, so it will hurt performance.

[No new Bamboo errors](https://lc.llnl.gov/bamboo/browse/LBANN-TIM353-1).

### CODE REVIEW

**Review by samadejacobs — APPROVED** [review]
LGTM
